### ISSUE TITLE
Rules per type of file

### ISSUE THREAD

**Issue by jogelin** [softarc-consulting/sheriff#19]
Hey,

I am accustomed to the boundaries and rules of Nx, and I really like your approach, which is more centralized and customizable!

I opened an issue on the Nx GitHub concerning the fact that I would like to set boundaries based on file extensions because I believe that a package can serve multiple goals such as components, specs, stories, etc...

You can find more information with the schema on https://github.com/nrwl/nx/issues/18151 .

Do you think your library can help with that?

What do you think about the idea of setting tags per file extension? If you like the idea, I can make a proposition and a PR ;)

**Comment by jogelin** [comment]
We could have different configuration files per extension like, wdyt?

**Comment by rainerhahnekamp** [comment]
Hi, that sounds very interesting. That could prevent somebody from mistakenly adding a story or spec file to production code. Additionally, we sometimes might need more "tolerant" dependency rules for test files.

That would mean that an `index.ts` or module can have different tags based on the extension, right?

What about removing the module encapsulation feature for file-based tagging? So we could say everything with *.spec, regardless of its directory, is tagged with `test`. Can you think of some use cases that might benefit more from that approach?

@manfredsteyer, @mikezks, anything to add here?

**Comment by jogelin** [comment]
> That could prevent somebody from mistakenly adding a story or spec file to production code.

It is exactly with one of the dev on my previous project -> fake library in prod :s

> That would mean that an index.ts or module can have different tags based on the extension, right?

Could be `index.ts` by default but with the possibility of specializing it

> What about removing the module encapsulation feature for file-based tagging? So we could say everything with *.spec, regardless of its directory, is tagged with test. Can you think of some use cases that might benefit more from that approach?

I still like considering the module/folder idea to scope multiple files and be able to specify scope tags more related to the domain/business/team/...

But a module/folder can serve multiple goals and I want to ensure that each goal is restricted to some shared library for example.

This is the other use cases I see
- ensure `.spec` file only imports test utilities + global utilities
- ensure `.component` only import global utilities 
- ensure `.stories` only import stories utilities + global utilities
- ensure `.cy` only import e2e utilities + global utilities
- ...

Related to that:
- I like the idea of grouping multiple types of files in one library because it is related to the same context.
- I don't like having one dumb graph that creates dependencies at the library level
- So I need to ensure linting helps me on that :)

**Comment by rainerhahnekamp** [comment]
Yeah, so all of that makes complete sense to me. I don't see any reason why Sheriff should support it. After #15 and #17 are done, this would be the next one.

**Comment by mikezks** [comment]
Great idea.

@rainerhahnekamp: as discussed, the Sheriff config could use the glob syntax and RexEx for advanced configs.

**Comment by rainerhahnekamp** [comment]
I hope that glob syntax will be enough. Regex is already possible, but when you cannot use a RegEx object as key in an object literal. And that is required for the config object. Glob can be a string as well.

So the first step would be to widen the definition of a module. In most cases, it will be the `index.ts` but we have to internally add the glob/regex expression as well...

**Comment by rainerhahnekamp** [comment]
@jogelin no progress yet on this feature, but it will eventually land. If you want (and have time), you can of course also come up with a PR

**Comment by rainerhahnekamp** [comment]
@jogelin, what about? 

```typescript
export const sheriffConfig: SheriffConfig = {
  tagging: {
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>-test'], selector: '**/*spec.ts},
    'src/app/<domain>/<type>': ['domain:<domain>', 'type:<type>'],
  },
  depRules: {
    // ...
  },
};
```

I would like that the keys stay as they are, and provide specialised configuration via the values. The same is true for the `depRules`.

Do you see something that speaks against this?

**Comment by jogelin** [comment]
@rainerhahnekamp it will work indeed!

When I saw that issue #36 and also your comment https://github.com/softarc-consulting/sheriff/issues/19#issuecomment-1665472745, I was thinking that maybe it could be solved by making the definition of the module more flexible.

- For the moment it is related to`/path/to/folder/index.ts` right?
- But on #36 you are proposing to have something configurable and provide any file name like`/path/to/folder/public-api.ts` right?
- I was thinking about having module definition more transversal and less focused on barrel files. Anything could be a module if it matches the pattern:
  - `src/app/<domain>/<type>/index.ts`
  - `src/app/**/index.ts`
  - `src/app/<domain>/<type>/public-api.ts`
  - `src/app/<domain>/<type>/**/*.spec.ts`
  - `src/app/<domain>/<type>/**/*-internal.ts`
  
  but could be solved in the same way like:
```js
export const sheriffConfig: SheriffConfig = {
  tagging: {
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>-test'], selector: '**/*spec.ts'},
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>'], selector: 'public-api.ts'},
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>'], selector: '**/*-internal.ts'},
  },
  depRules: {
    // ...
  },
};
```
I just don't see the added value of splitting the path and the selectors.

I would see that more like [Nx project inference](https://nx.dev/extending-nx/recipes/project-graph-plugins#identifying-projects), everything that matches a path could be considered as an Nx Project.

wdyt?

**Comment by rainerhahnekamp** [comment]
Well, the question is always who defines which elements the module exposes. That's why we have a barrel file.

In nx that works because a project can be used for many reasons. Sheriff shouldn't be Nx. It should focus on the module part.

Depending in which direction (barrel file a problem because of tree-shaking) we go:

**with barrel.file**
```typescript
export const sheriffConfig: SheriffConfig = {
  tagging: {
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>-test'], selector: '**/*spec.ts', barrel: 'index.ts'},
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>'], selector: '**/*-internal.ts', barrel: 'internal.index.ts'},
  },
  depRules: {
    // ...
  },
};
```

If `barrel` is `index.ts` we could actually skip it and use it as default.

**without barrel**

```typescript
export const sheriffConfig: SheriffConfig = {
  tagging: {
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>-test'], selector: '**/*spec.ts', barrel: false},
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>'], selector: '**/*-internal.ts', barrel: false},
  },
  depRules: {
    // ...
  },
};
```

Here, we have a problem. Without barrel, we could say all files in the directory are exposed and all within subfolders are hidden. But now we have one directory hosting two modules.

**Comment by rainerhahnekamp** [comment]
@chwoerz: maybe you want to add something to this discussion

**Comment by jogelin** [comment]
@rainerhahnekamp I think I understand where I am wrong.

Actually, I am not considering that boundaries should be applied only on files exposed by a public API, a barrel file. 

I was considering a module as a group of files without any public entry point.

My reflexion is more oriented by "Can this **file or list of files** can access to this **file or list of files**". 
And then adding a layer on top to recognize barrel files as a way to get list of files.

**Comment by rainerhahnekamp** [comment]
I understand. So you have a module without encapsulation but with dependency rules.

I would see this as a special case which is something that can be supported. We have to ensure that necessary changes to the config format has no negative effects (more boilerplaty) for the default use cases.

In a way, we could go with the configuration posted above and `barrel: false` just means that it is a module without encapsulation.

```typescript
export const sheriffConfig: SheriffConfig = {
  tagging: {
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>-test'], selector: '**/*spec.ts', barrel: false},
    'src/app/<domain>/<type>': {tags: ['domain:<domain>', 'type:<type>'], selector: '**/*-internal.ts', barrel: false},
  },
  depRules: {
    // ...
  },
};
```

Does this look reasonable?

In the meantime, let me think about potential side effects if a module has no barrel file or `_private` directory (#47 )
